crates/staking/README.md

```diff
@@ -0,0 +1,131 @@
+
+We present the simple solution (without rewards) below:
+
+```python
+class Slashing:
+    def __init__(self):
+        self.total_stake = 0
+        self.slash_per_token = 0
+        self.stake = {}
+        self.slash_tally = {}
+
+    def deposit_stake(self, address, amount):
+        if address not in self.stake:
+            self.stake[address] = 0
+            self.slash_tally[address] = 0
+
+        self.stake[address] += amount
+        self.slash_tally[address] += self.slash_per_token * amount
+        self.total_stake += amount
+
+    def slash_stake(self, amount):
+        self.slash_per_token += amount / self.total_stake
+
+    def compute_stake(self, address):
+        to_slash = self.stake[address] * self.slash_per_token - self.slash_tally[address]
+        return self.stake[address] - to_slash
+
+    def withdraw_stake(self, address, amount):
+        if amount > self.compute_stake(address):
+            raise Exception("Invalid amount")
+
+        self.stake[address] -= amount
+        self.slash_tally[address] -= self.slash_per_token * amount
+        self.total_stake -= - amount
+```
+
+Without accounting for changing stake sizes we can reduce the staking calculation to the following formula:
+
+```
+actual_stake = stake - (stake * (slashed_amount / total_stake))
+```
+
+### Example
+
+1. Alice deposits 100 DOT in staking pool A.
+2. Bob deposits 50 DOT in staking pool A.
+
+At this point in time, assuming we slash 50 DOT in A, we have the following stakes:
+
+- Alice may withdraw `100 - (100 * (50 / 150)) = 66.66 DOT`.
+- Bob may withdraw `50 - (50 * (50 / 150)) = 33.33 DOT`.
+
+### Unslashing
+
+It is reasonable to assume that stake may be re-distributed to participants after slashing. In this case we would want to "unslash" participant stake proportionally. Since we calculate the amount to slash dynamically we can handle this case by reducing `slash_per_token` by the amount to release. An example of where this might be useful is if the Vault is liquidated with open redeem requests, upon completion we should re-distribute collateral to the Vault and its Nominators.
+
+## Rewards
+
+The proportionality of rewards is directly tied to the stake of a participant. Assuming stake may be slashed after reward distribution and prior to reward withdrawal, we must make slight adjustments to the [scalable reward distribution](https://solmaz.io/2019/02/24/scalable-reward-changing/). For instance, upon slashing we increase `reward_per_token` by `(amount * reward_per_token) / total_stake` to account for lost rewards. Additionally, we must re-compute stake in `withdraw_rewards` to get the correct balance for an account.
```

Comment:
It seems odd to me that you talk about slashing, and then proceed to present the reward calculation that does not account for slashing.

Reply:
We have two options, combine the python scripts or remove them altogether (and maybe link to the econ repo). What do you think?